[serve] Reduce surface area of `create_router` interface

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Moves more of the initialization logic into `create_router` so `DeploymentHandle` does not depend on being run inside an initialized Ray context.

This is required so I can use the same `DeploymentHandle` implementation for local testing mode and only inject a different `create_router` implementation.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

zcin commented on python/ray/serve/handle.py:
```diff
@@ -175,24 +149,9 @@ def _set_request_protocol(self, request_protocol: RequestProtocol):
             self._router = create_router(
-                serve.context._get_global_client()._controller,
-                self.deployment_id,
                 self.handle_id,
```
```suggestion
                handle_id=self.handle_id,
```

zcin commented on python/ray/serve/handle.py:
```diff
@@ -175,24 +149,9 @@ def _set_request_protocol(self, request_protocol: RequestProtocol):
-                get_current_actor_id(),
-                availability_zone,
-                _create_or_get_global_asyncio_event_loop_in_thread(),
+                self.deployment_id,
```
```suggestion
                deployment_id=self.deployment_id,
```

zcin commented on python/ray/serve/_private/default_impl.py:
```diff
@@ -67,16 +69,55 @@ def create_init_handle_options(**kwargs):
     return _InitHandleOptions.create(**kwargs)
 
 
+_global_async_loop = None
```
Should this be renamed to `_global_async_router_loop` or something similar, to indicate it's the loop for the router?

edoakes replied:
yea good idea

edoakes commented:
@zcin please take another look. I updated the code to add a `Router` abstract base class. The interface for this should return a `concurrent.futures.Future`, so I've renamed the existing router class to `AsyncioRouter` and wrapped it in a `SingletonThreadRouter` which now manages the global singleton thread logic (much cleaner IMO).

zcin commented on python/ray/serve/_private/router.py:
```diff
@@ -560,7 +584,65 @@ async def assign_request(
+        **request_kwargs,
+    ) -> concurrent.futures.Future[ReplicaResult]:
+        return asyncio.run_coroutine_threadsafe(
+            self._asyncio_router.assign_request(
+                request_meta, *request_args, **request_kwargs
+            ),
+            loop=self._asyncio_loop,
+        )
+
     def shutdown(self):
         asyncio.run_coroutine_threadsafe(
-            self._metrics_manager.shutdown(), loop=self._event_loop
+            self._asyncio_router.shutdown(), loop=self._asyncio_loop
         ).result()
```
Ah this is so much nicer

zcin commented on python/ray/serve/_private/default_impl.py:
```diff
@@ -98,20 +117,22 @@ def create_router(
+            handle_id=handle_id,
+            self_actor_id=actor_id,
+            handle_source=handle_options._source,
+            event_loop=event_loop,
```
Hmm getting the event loop from a class method on `SingletonThreadRouter`, then feeding it into `AsyncioRouter` for initialization, then feeding that back into `SingletonThreadRouter` seems a bit confusing to me. Would it make more sense for `SingletonThreadRouter` to instantiate `AsyncioRouter` instead of an instance of `AsyncioRouter` being passed in?

edoakes replied:
Yeah I agree that part is a bit funky... the reason I didn't do what you suggested initially was to avoid duplicating all of the init args. But I'm now realizing I could just leave them as passthrough kwargs -- let me do that

edoakes replied:
updated

zcin commented on python/ray/serve/handle.py:
```diff
@@ -134,6 +108,7 @@ def __init__(
         *,
         handle_options: Optional[_DynamicHandleOptionsBase] = None,
         _router: Optional[Router] = None,
+        _create_router: Optional[CreateRouterCallable] = None,
```
Is this going to be used for local testing mode?

edoakes replied:
yes

zcin reviewed (approved):
LGTM!
